subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -1154,3 +1158,143 @@ int nrf_cloud_rest_jitp(const sec_tag_t nrf_cloud_sec_tag)
 
 	return ret;
 }
+
+int nrf_cloud_rest_send_location(struct nrf_cloud_rest_context *const rest_ctx,
+	const char *const device_id, const char *const nmea_sentence)
+{
+	__ASSERT_NO_MSG(rest_ctx != NULL);
+	__ASSERT_NO_MSG(device_id != NULL);
+	__ASSERT_NO_MSG(nmea_sentence != NULL);
+
+	int err = -ENOMEM;
+	char *json_msg = NULL;
+
+	(void)nrf_cloud_codec_init();
+
+	cJSON *root_obj = json_create_req_obj(NRF_CLOUD_JSON_APPID_VAL_GPS,
+					     NRF_CLOUD_JSON_MSG_TYPE_VAL_DATA);
+
+	if (cJSON_AddStringToObjectCS(root_obj, NRF_CLOUD_JSON_DATA_KEY, nmea_sentence) == NULL) {
+		LOG_ERR("Failed to add NMEA string");
+		goto clean_up;
+	}
+
+	json_msg = cJSON_PrintUnformatted(root_obj);
+	if (!json_msg) {
+		LOG_ERR("Failed to print JSON");
+		goto clean_up;
+	}
+
+	err = nrf_cloud_rest_send_device_message(rest_ctx, device_id, json_msg, NULL);
+
+clean_up:
+	cJSON_Delete(root_obj);
+	if (json_msg) {
+		cJSON_free((void *)json_msg);
+	}
+
+	return err;
+}
+
+int nrf_cloud_rest_send_device_message(struct nrf_cloud_rest_context *const rest_ctx,
+	const char *const device_id, const char *const json_msg, const char *const topic)
+{
+	__ASSERT_NO_MSG(rest_ctx != NULL);
+	__ASSERT_NO_MSG(device_id != NULL);
+	__ASSERT_NO_MSG(json_msg != NULL);
+
+	int ret;
+	size_t buff_sz;
+	char *auth_hdr = NULL;
+	char *url = NULL;
+	cJSON *root_obj;
+	cJSON *msg_obj;
+	struct rest_client_req_context req;
+	struct rest_client_resp_context resp;
+
+	memset(&resp, 0, sizeof(resp));
+	init_rest_client_request(rest_ctx, &req, HTTP_POST);
+
+	root_obj = cJSON_CreateObject();
+	msg_obj = cJSON_Parse(json_msg);
+
+	if (cJSON_AddItemToObjectCS(root_obj, API_DEVICES_MSGS_MSG_KEY, msg_obj) != cJSON_True) {
+		ret = -ENOMEM;
+		cJSON_Delete(msg_obj);
+		goto clean_up;
+	}
+
+	if (topic) {
+		if (cJSON_AddStringToObjectCS(root_obj, API_DEVICES_MSGS_TOPIC_KEY,
+					      topic) == NULL) {
+			ret = -ENOMEM;
+			goto clean_up;
+		}
+	}
+
+	/* Set payload */
+	req.body = cJSON_PrintUnformatted(root_obj);
+	if (!req.body) {
+		ret = -ENOMEM;
+		goto clean_up;
+	}
+
+	/* Format API URL with device ID */
+	buff_sz = sizeof(API_DEVICES_MSGS_TEMPLATE) + strlen(device_id);
+	url = k_calloc(buff_sz, 1);
+	if (!url) {
+		ret = -ENOMEM;
+		goto clean_up;
+	}
+	req.url = url;
+
+	ret = snprintk(url, buff_sz, API_DEVICES_MSGS_TEMPLATE, device_id);
+	if (ret < 0 || ret >= buff_sz) {
+		LOG_ERR("Could not format URL");
+		ret = -ETXTBSY;
+		goto clean_up;
+	}
+
+	/* Format auth header */
+	ret = generate_auth_header(rest_ctx->auth, &auth_hdr);
+	if (ret) {
+		LOG_ERR("Could not format HTTP auth header");
+		goto clean_up;
+	}
+	char *const headers[] = {
```

Comment:
Nit: This can be moved to the top of the function with the other variable definitions

Reply:
`auth_hdr` is initially NULL and allocated just a couple lines above in `generate_auth_header`.